server/server_web_push.go

```diff
@@ -30,109 +30,98 @@ var webPushEndpointAllowRegex = regexp.MustCompile(webPushEndpointAllowRegexStr)
 
 func (s *Server) handleWebPushUpdate(w http.ResponseWriter, r *http.Request, v *visitor) error {
 	payload, err := readJSONWithLimit[webPushSubscriptionPayload](r.Body, jsonBodyBytesLimit, false)
-
 	if err != nil || payload.BrowserSubscription.Endpoint == "" || payload.BrowserSubscription.Keys.P256dh == "" || payload.BrowserSubscription.Keys.Auth == "" {
 		return errHTTPBadRequestWebPushSubscriptionInvalid
-	}
-
-	if !webPushEndpointAllowRegex.MatchString(payload.BrowserSubscription.Endpoint) {
+	} else if !webPushEndpointAllowRegex.MatchString(payload.BrowserSubscription.Endpoint) {
 		return errHTTPBadRequestWebPushEndpointUnknown
-	}
-
-	if len(payload.Topics) > webPushTopicSubscribeLimit {
+	} else if len(payload.Topics) > webPushTopicSubscribeLimit {
 		return errHTTPBadRequestWebPushTopicCountTooHigh
 	}
-
-	u := v.User()
-
 	topics, err := s.topicsFromIDs(payload.Topics...)
 	if err != nil {
 		return err
 	}
-
 	if s.userManager != nil {
+		u := v.User()
 		for _, t := range topics {
 			if err := s.userManager.Authorize(u, t.ID, user.PermissionRead); err != nil {
 				logvr(v, r).With(t).Err(err).Debug("Access to topic %s not authorized", t.ID)
 				return errHTTPForbidden.With(t)
 			}
 		}
 	}
-
 	if err := s.webPush.UpdateSubscriptions(payload.Topics, v.MaybeUserID(), payload.BrowserSubscription); err != nil {
 		return err
 	}
-
 	return s.writeJSON(w, newSuccessResponse())
 }
 
 func (s *Server) publishToWebPushEndpoints(v *visitor, m *message) {
 	subscriptions, err := s.webPush.SubscriptionsForTopic(m.Topic)
 	if err != nil {
 		logvm(v, m).Err(err).Warn("Unable to publish web push messages")
-
 		return
 	}
-
-	for i, xi := range subscriptions {
-		go func(i int, sub webPushSubscription) {
-			ctx := log.Context{"endpoint": sub.BrowserSubscription.Endpoint, "username": sub.UserID, "topic": m.Topic, "message_id": m.ID}
-
-			s.sendWebPushNotification(newWebPushPayload(fmt.Sprintf("%s/%s", s.config.BaseURL, m.Topic), *m), &sub, &ctx)
-		}(i, xi)
+	payload, err := json.Marshal(newWebPushPayload(fmt.Sprintf("%s/%s", s.config.BaseURL, m.Topic), m))
+	if err != nil {
+		log.Tag(tagWebPush).Err(err).Warn("Unable to marshal expiring payload")
+		return
+	}
+	for _, subscription := range subscriptions {
+		ctx := log.Context{"endpoint": subscription.BrowserSubscription.Endpoint, "username": subscription.UserID, "topic": m.Topic, "message_id": m.ID}
+		if err := s.sendWebPushNotification(payload, subscription, &ctx); err != nil {
+			log.Tag(tagWebPush).Err(err).Fields(ctx).Warn("Unable to publish web push message")
+		}
```

Comment:
Doesn't this run sequentially now, without the `go func`?

Reply:
It does, but the whole thing runs in a go routine. No need to spin up more and make the code less readable IMHO.

Reply:
fair enough, I'm just imagining if there were hundreds of web push subscriptions on a topic, this would take some time to send them all, vs running all web push sends in parallel. and it would also make the publisher wait, I believe?

Reply:
The `publishToWebPushEndpoints` function is called in two places, and it's only called as `go publishToWebPushEndpoints()`, so the publisher doesn't have to wait. And usually there are only 1-2 subscribers. There are only a handful of topics with more subscribers.